features

MNowicki87 opened the pull request:
I know the code is ugly as hell and have no intention of making it better any time soon :-P
Home page displays a form to choose reservation dates and branches, posting form returns a list of sample cars (by category) that are currently available for set time period in selected pickup branch.

MNowicki87 committed:
the 1000 cars now are categorized

MNowicki87 committed:
Merge branch 'master' into local_mn

MNowicki87 committed:
- homePage includes reservation form
- unplugged 'mdboostrap'
- added messages.properties
- styled some FE

MNowicki87 committed:
- can list car categories available for rent in specified branch and dates
- moved Car -> CarCategory for low count notifications and everything reservation related (actual hire and return are still car-specific
- added methods to CarService
- moved static final variabled to DomainValues class
-

MNowicki87 committed:
Merge branch 'master' into develop_MN

Mbojanow commented on src/main/java/pl/sdacademy/carrental/configuration/DomainValues.java:
```diff
@@ -0,0 +1,11 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
definately not a component :)

MNowicki87 replied:
indeed it is not (well, not anymore) :-)

Mbojanow commented on src/main/java/pl/sdacademy/carrental/configuration/DomainValues.java:
```diff
@@ -0,0 +1,11 @@
+
+@Component
+// TODO: 18/06/2020 it would probably be better to store these values in DB
+public class DomainValues {
```
if you want to use class for consts -> make it final. Other approach is to define those in an interface

MNowicki87 replied:
good point! done

Mbojanow commented on src/main/java/pl/sdacademy/carrental/controllers/BranchController.java:
```diff
@@ -21,9 +23,9 @@ public BranchController(final BranchService branchService) {
    @GetMapping("/list-branches")
    public String listAllBranches(final Model model) {
-      final Map<Branch, Long> branchesWithCarCount = branchService.getBranchesWithCarCount();
+      final Map<Branch, Map<CarCategory, Long>> branchesWithCarCount = branchService.getBranchesWithCarCount();
```
see Table from Guava that wraps Maps na use Map as a value

MNowicki87 replied:
tried it… iterating through the table and getting necessary info seemed like a nightmare

Mbojanow commented on src/main/java/pl/sdacademy/carrental/controllers/ReservationsController.java:
```diff
@@ -0,0 +1,37 @@
+      model.addAttribute("reservationRequest", reservationRequest);
+      model.addAttribute("carList", upForReservation);
+   
+      System.out.println("upForReservation = " + upForReservation);
```
use log instead of souts

MNowicki87 replied:
that was not supposed to be there - a leftover from debugging

Mbojanow commented on src/main/java/pl/sdacademy/carrental/controllers/ReservationsController.java:
```diff
@@ -0,0 +1,37 @@
+   }
+   
+   @PostMapping
+   public String showPageWithAvailableCarsGivenDateAndBranch(@ModelAttribute final CarReservationRequest reservationRequest,
```
not sure but no validation here?

MNowicki87 replied:
validation added

Mbojanow commented on src/main/java/pl/sdacademy/carrental/domain/rentals/Reservation.java:
```diff
@@ -57,8 +57,8 @@
-   @AssertTrue
-   private boolean returnDateShouldBeAfterPickupDate() {
+   @AssertTrue(message = "Return date must be after pickup date")
+   private boolean isReturnAfterPickup() {
```
might add nonull check maybe?

MNowicki87 replied:
validation added to entity class as well

Mbojanow commented on src/main/java/pl/sdacademy/carrental/repositories/BranchRepository.java:
```diff
@@ -13,4 +13,6 @@
 
     List<Branch> getAllByStatus(final BranchStatus status);
+   
+   Optional<Branch> findBranchByName(String branchName);
```
is the method name correct? findByName?

Mbojanow commented on src/main/java/pl/sdacademy/carrental/services/BranchService.java:
```diff
@@ -87,18 +88,26 @@ public BranchForm getById(final Long id) {
-                  .equals(Status.IN)).count();
+   private Map<CarCategory, Long> getAvailableCarCountPerCategory(final Branch branch) {
+       return branch.getCarsOnHand().stream()
+             .filter(car -> car.getCurrentStatus().equals(Status.IN))
```
enum can be checked with ==

MNowicki87 replied:
done

Mbojanow commented on src/main/java/pl/sdacademy/carrental/services/ReservationService.java:
```diff
@@ -0,0 +1,101 @@
+      
+      Map<CarCategory, Integer> numberOfAvailableCarsPerCategoryInPeriod = new HashMap<>();
+      
+      Arrays.asList(CarCategory.values()).forEach(category -> {
```
Stream.of(CarCategory.values()).forEach() and move lambda inside to a separate method or a separate bean

MNowicki87 replied:
Done, although I'm not sure about benefits of this change (`Stream.of(…)forEach(…)` over `Arrays.asList(…).forEach(…)`)

Mbojanow commented on src/main/resources/static/scripts/scripts.js:
```diff
@@ -0,0 +1,42 @@
+let navbar = document.querySelector("nav.navbar");
```
this navbar can be const

MNowicki87 replied:
yup, thanks -> replaced some lets with consts

Mbojanow commented on src/main/resources/static/scripts/scripts.js:
```diff
@@ -0,0 +1,42 @@
+    setReturnDateToMatchPickup();
+}
+
+let pickupDate = document.getElementById('pickup_date');
```
this can also be const?

MNowicki87 committed:
- revised links to homepage in navbar
- added a tooltip informing about possibility of returning a car to another branch and extra charge

MNowicki87 committed:
- switched to interface to store Domain constants
- switched to let to const wherever possible in scripts.js
- refactored ReservationService a bit

MNowicki87 committed:
- DomainValues -> back to class (final)

MNowicki87 committed:
- added 'Valid' to postMapping in ReservationsController
- fixed a minor bug in branches-list.html

MNowicki87 committed:
- changed enum comparison from equals() to '=='
- added validation for Reservation and CarReservationRequest

MNowicki87 committed:
Merge branch 'develop' into develop_MN

# Conflicts:
#	src/main/java/pl/sdacademy/carrental/services/BranchService.java

MNowicki87 committed:
branchService.findBranchByName throws a BranchException

MNowicki87 committed:
Client entity -> replaced regex validation with Digits annotation)

MNowicki87 committed:
Merge branch 'master' into develop_MN

MNowicki87 committed:
- added head tags to views
- fixed links to stylesheets and scripts
